src/components/TableOfContents.astro

```diff
@@ -0,0 +1,125 @@
+---
+
+<nav id="table-of-contents" aria-label="Table Of Contents">
+    <ol>
+        {
+            grouppedHeadings.map(h => (
+                <li>
+                    <a href={`#${h.slug}`}>{h.text}</a>
+                    {h.subheadings.length > 0 && (
+                        <ol>
+                            {h.subheadings.map(sub => (
+                                <li>
+                                    <a href={`#${sub.slug}`}>{sub.text}</a>
+                                </li>
+                            ))}
+                        </ol>
+                    )}
+                </li>
+            ))
+        }
+    </ol>
+</nav>
+
+<script is:inline>
+    // This script tag is useful only if you want to display the TOC alongside the blog post...
+    // ... and highlight the section that the user is currently reading through.
+    // Feel free to remove this tag if you don't need this type of functionality.
+
+    const observer = new IntersectionObserver(
+        entries => {
+            for (const entry of entries) {
+                const headingFragment = `#${entry.target.id}`;
+                const tocItem = document.querySelector(`a[href="${headingFragment}"]`);
+
+                if (entry.isIntersecting) {
+                    const previouslyActivatedItem = document.querySelector(".active-toc-item");
+
+                    previouslyActivatedItem?.classList.remove("active-toc-item");
+                    tocItem.classList.add("active-toc-item");
+                } else {
+                    const isAnyOtherEntryIntersecting = entries.some(
+                        e => e.target.id !== entry.target.id && e.isIntersecting,
+                    );
+
+                    if (isAnyOtherEntryIntersecting) {
+                        tocItem.classList.remove("active-toc-item");
+                    }
+                }
+            }
+        },
+        { root: null, rootMargin: "0px", threshold: [1] },
+    );
+
+    const sectionHeadings = document.querySelectorAll("article > h2, article > h3");
+
+    for (const heading of sectionHeadings) {
+        observer.observe(heading);
+    }
+
+    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
+        anchor.addEventListener("click", function (e) {
+            e.preventDefault();
+
+            document.querySelector(this.getAttribute("href")).scrollIntoView({
+                behavior: "smooth",
+            });
+        });
+    });
+</script>
+
+<style>
+    .active-toc-item {
+        font-weight: bold;
+    }
+
+    #table-of-contents {
+        font-family: Arial, sans-serif;
+        font-size: 16px;
+    }
+
+    #table-of-contents > ol {
+        padding-left: 0;
+        margin: 0;
+    }
+
+    #table-of-contents > ol > li {
+        margin-bottom: 0.5em;
+    }
+
+    #table-of-contents ol ol {
+        padding-left: 20px;
+        margin: 0;
+    }
+
+    #table-of-contents ol ol li {
+        margin-bottom: 0.5em;
+    }
+</style>
```

Comment:
Wondering if there are more elegant solutions to this. Is this an original work? Curious to see the origin if so, or to hear why you did this in such a way.

Reply:
nah I took the toc component from another GitHub repo and asked Copilot to add spaces between sub-headers. If have refactored the section to make it a bit better

Reply:
I'll refactor this separately if at all possible.